rspace/src/main/scala/coop/rchain/rspace/merger/StateChange.scala

```diff
@@ -0,0 +1,159 @@
+
+final case class StateChange(
+    datumChanges: Map[Blake2b256Hash, ChannelChange[ByteVector]],
+    kontChanges: Map[Blake2b256Hash, ChannelChange[ConsumeChange]],
+    joinsIndex: Set[JoinIndex]
+)
+
+object StateChange {
+
+  private def computeValueChangeOnChannel[F[_]: Sync, T](
+      valuePointer: Blake2b256Hash,
+      getStartValue: Blake2b256Hash => F[Seq[T]],
+      getEndValue: Blake2b256Hash => F[Seq[T]]
+  ): F[ChannelChange[T]] =
+    for {
+      startValue <- getStartValue(valuePointer)
+      endValue   <- getEndValue(valuePointer)
+      added      = endValue diff startValue
+      deleted    = startValue diff endValue
+    } yield ChannelChange(added.toVector, deleted.toVector)
+
+  private def recordValueChange[F[_]: Concurrent, T](
+      valuePointer: Blake2b256Hash,
+      startValueReader: Blake2b256Hash => F[Seq[T]],
+      endValueReader: Blake2b256Hash => F[Seq[T]],
+      accumulator: Ref[F, Map[Blake2b256Hash, ChannelChange[T]]]
+  ): Stream[F, Unit] =
+    fs2.Stream.eval(for {
+      r <- computeValueChangeOnChannel[F, T](
+            valuePointer,
+            h => startValueReader(h),
+            h => endValueReader(h)
+          )
+      _ <- accumulator.update(s => {
+            val curVal = s.getOrElse(valuePointer, ChannelChange.empty)
+            val newVal = curVal.copy(
+              added = curVal.added ++ r.added,
+              removed = curVal.removed ++ r.removed
+            )
+            s.updated(valuePointer, newVal)
+          })
+    } yield ())
+
+  def apply[F[_]: Concurrent, C, P, A, K](
+      preStateReader: HistoryReaderBinary[F, C, P, A, K],
+      postStateReader: HistoryReaderBinary[F, C, P, A, K],
+      eventLogIndex: EventLogIndex,
+      channelsStore: ChannelStore[F, C],
+      serializeC: Serialize[C]
+  ): F[StateChange] =
+    for {
+      // accumulators for data required to create history actions
+      // changes to produces, to consumes and cumulative list of joins
+      producesDiffRef <- Ref.of[F, Map[Blake2b256Hash, ChannelChange[ByteVector]]](Map.empty)
+      consumesDiffRef <- Ref.of[F, Map[Blake2b256Hash, ChannelChange[ConsumeChange]]](Map.empty)
+      joinsIndexRef   <- Ref.of[F, Set[JoinIndex]](Set.empty)
+
+      // TODO remove when channels map is removed
+      producePointers <- channelsStore.getProduceMappings(producesAffected(eventLogIndex).toSeq)
+      consumePointers <- channelsStore.getConsumeMappings(consumesAffected(eventLogIndex).toSeq)
+      joinsPointers <- channelsStore.getJoinMapping(
+                        (eventLogIndex.consumesLinearAndPeeks ++ eventLogIndex.consumesPersistent ++ eventLogIndex.consumesProduced).toSeq
+                          .flatMap(_.channelsHashes)
+                      )
+
+      // `distinct` required here to not compute change for the same hash several times
+      // as channel can be mentioned several times in event log
+      affectedDatumsValuePointers = producePointers.map(_.historyHash).distinct
```

Comment:
Maybe it is faster to distinct first and then read the mapping although it is not important in the future that we don't have channelStore.